service/history/taskPriorityAssigner.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"sync"
+
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/quotas"
+)
+
+type (
+	taskPriorityAssigner interface {
+		Assign(queueTask) error
+	}
+
+	taskPriorityAssignerImpl struct {
+		currentClusterName string
+		domainCache        cache.DomainCache
+		config             *Config
+		logger             log.Logger
+		scope              metrics.Scope
+
+		sync.RWMutex
```

Comment:
It's a common practice to declare anonymous mutex as the first member of the struct.

Reply:
Sure I can update. But I think in our code base we are following two different conventions: one is but the mutex as the first member of the struct as you suggested. The other one is put it just above the field the lock is protecting. Examples are `queueAckMgr.go`, `shardContext.go` etc.

Reply:
For the latter, anonymous mutexes don't make sense since the mutex is only protecting a sub-resource within the struct. For these cases, the pattern should to be use a named mutex variable. For example, if mutex protects some resourceMap, then mutex should be named resourceMapMutex.

Reply:
Got it! Thanks for explaining.